rs/index/src/ivf/builder.rs

```diff
@@ -68,29 +77,47 @@ impl IvfBuilder {
         Ok(())
     }
 
-    pub fn build_inverted_lists(
-        vector_storage: &FixedFileVectorStorage<f32>,
-        centroid_storage: &FixedFileVectorStorage<f32>,
-    ) -> Result<HashMap<usize, Vec<usize>>> {
-        let mut context = SearchContext::new(false);
-        let mut inverted_lists = HashMap::new();
+    /// Add a posting list
+    pub fn add_posting_list(&mut self, posting_list: &[u64]) -> Result<()> {
+        self.posting_lists.append(posting_list)?;
+        Ok(())
+    }
+
+    fn find_nearest_centroids(
+        vector: &[f32],
+        centroids: &dyn VectorStorage<f32>,
+        num_probes: usize,
+    ) -> Result<Vec<usize>> {
+        let mut distances: Vec<(usize, f32)> = Vec::new();
+        for i in 0..centroids.len() {
+            let centroid = centroids.get(i as u32)?;
+            let dist = L2DistanceCalculator::calculate(&vector, &centroid);
+            distances.push((i, dist));
+        }
+        distances.select_nth_unstable_by(num_probes - 1, |a, b| a.1.total_cmp(&b.1));
+        Ok(distances.into_iter().map(|(idx, _)| idx).collect())
+    }
+
+    pub fn build_posting_lists(&mut self) -> Result<()> {
+        let mut posting_lists: Vec<Vec<u64>> =
+            vec![Vec::with_capacity(0); self.config.num_clusters as usize];
         // Assign vectors to nearest centroid
-        for i in 0..vector_storage.num_vectors {
-            let vector = vector_storage
-                .get(i, &mut context)
-                .ok_or(anyhow::anyhow!("Failed to get vector at index {}", i))?
-                .to_vec();
-            let nearest_centroid = Ivf::find_nearest_centroids(&vector, &centroid_storage, 1)?;
-            inverted_lists
-                .entry(nearest_centroid[0])
-                .or_insert_with(Vec::new)
-                .push(i);
+        for i in 0..self.vectors.len() {
+            let vector = self.vectors.get(i as u32)?;
+            let nearest_centroid =
+                Self::find_nearest_centroids(&vector, self.centroids.as_ref(), 1)?;
```

Comment:
interestingly, a document _could_ be in multiple centroids. let's add a config saying how many centroids a document could be in (default to 1), and then take that many nearest centroids and add it to the posting list